rust/src/builder.rs

```diff
@@ -86,29 +186,19 @@ where
     /// Grow the internal buffer to `new_capacity`, where `new_capacity` is the capacity in
     /// elements of type T
     fn grow(&mut self, new_capacity: i64) -> Result<()> {
-        let byte_capacity = mem::size_of::<T>() * new_capacity as usize;
-        self.buffer.resize(byte_capacity)
+        let new_byte_capacity = bit_util::ceil(new_capacity, 8) as usize;
+        let existing_capacity = self.buffer.capacity();
+        let capacity_added = self.buffer.resize(new_byte_capacity)?;
+        self.buffer
+            .ensure_null_bits(existing_capacity, capacity_added);
+        Ok(())
     }
 
     /// Build an immutable `Buffer` from the existing internal `MutableBuffer`'s memory
     pub fn finish(self) -> Buffer {
-        self.buffer.freeze()
-    }
-
-    /// Writes a byte slice to the underlying buffer and updates the `len`, i.e. the number array
-    /// elements in the builder.  Also, converts the `io::Result` required by the `Write` trait
-    /// to the Arrow `Result` type.
-    fn write_bytes(&mut self, bytes: &[u8], len_added: i64) -> Result<()> {
```

Comment:
Do we need to implement `write_bytes` for boolean type?

Reply:
I only needed `write_bytes` previously to write a zero byte as I pushed on new elements.  We are now ensuring that the entire capacity is zero when we create or resize so we don't need this anymore.